File: Binary Tree Postorder Traversal.cpp
/*
    Binary Tree Postorder Traversal Total Accepted: 12510 Total Submissions: 41534 My Submissions
Given a binary tree, return the postorder traversal of its nodes' values.

For example:
Given binary tree {1,#,2,3},
   1
    \
     2
    /
   3
return [3,2,1].

Note: Recursive solution is trivial, could you do it iteratively?
*/
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode *root) {
        vector<int> result;
        if(root==NULL)
            return result;
        postorderHelper(root,result);
        return result;
    }
    void postorderHelper(TreeNode *root, vector<int>&result){
        if(root==NULL)
            return;
        if(root->left!=NULL){
            postorderHelper(root->left,result);
        }
        if(root->right!=NULL){
            postorderHelper(root->right,result);
        }
        result.push_back(root->val);
    }
};

File: Template/Binary_Tree_Postorder_Traversal.java
/*
Binary Tree Postorder Traversal Total Accepted: 25109 Total Submissions: 81165 My Submissions
Given a binary tree, return the postorder traversal of its nodes' values.

For example:
Given binary tree {1,#,2,3},
   1
    \
     2
    /
   3
return [3,2,1].
*/
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<Integer>();
        if (root == null) {
            return result;
        }
        
        Stack<TreeNode> s = new Stack<TreeNode>();
        TreeNode current = root;
        while (true) {
            while (current != null) {
                if (current.right != null) {
                    s.push(current.right);
                }
                s.push(current);
                current = current.left;
            }
        
            if (s.isEmpty()) {
                return result;
            }
            
            current = s.pop();
            if (current.right != null && !s.isEmpty() && current.right == s.peek()) {
                s.pop();
                s.push(current);
                current = current.right;
            }
            else {
                result.add(current.val);
                current = null;
            }
        }
    }
}